lib/SILOptimizer/Utils/CanonicalizeOSSALifetime.cpp

```diff
@@ -611,6 +614,121 @@ void CanonicalizeOSSALifetime::insertDestroysOnBoundary(
+  auto currentDef = getCurrentDef();
+
+  // First, collect the blocks that were _originally_ live.  We can't use
+  // liveness here because it doesn't include blocks that occur before a
+  // destroy_value.
+  BasicBlockSet originalLiveBlocks(currentDef->getFunction());
+  {
+    // Start the walk from the consuming blocks (which includes destroys as well
+    // as the other consuming uses).
+    BasicBlockWorklist worklist(currentDef->getFunction());
+    for (auto *consumingBlock : consumingBlocks) {
+      worklist.push(consumingBlock);
+    }
+
+    // Walk backwards from consuming blocks.
+    auto *currentDefBlock = currentDef->getParentBlock();
+    while (auto *block = worklist.pop()) {
+      originalLiveBlocks.insert(block);
+      // Keep adding blocks until we find the def block.
+      if (currentDefBlock == block)
+        continue;
+      for (auto *predecessor : block->getPredecessorBlocks()) {
+        worklist.pushIfNotVisited(predecessor);
+      }
+    }
+  }
+
+  // Second, collect the blocks which occur after a _final_ consuming use.
+  BasicBlockSetVector finalConsumingBlocks(currentDef->getFunction());
+  BasicBlockSetVector consumedBlocks(currentDef->getFunction());
+  {
+    // Start the walk from blocks which contain _final_ non-destroy consumes.
+    // These are just the instructions on the boundary which aren't destroys.
+    BasicBlockWorklist worklist(currentDef->getFunction());
+    for (auto *instruction : boundary.lastUsers) {
+      if (dynCastToDestroyOf(instruction, getCurrentDef()))
+        continue;
+      if (liveness.isInterestingUser(instruction) !=
+          PrunedLiveness::IsInterestingUser::LifetimeEndingUse)
+        continue;
+      worklist.push(instruction->getParent());
+    }
+    while (auto *block = worklist.pop()) {
+      finalConsumingBlocks.insert(block);
+      for (auto *successor : block->getSuccessorBlocks()) {
+        if (!originalLiveBlocks.contains(successor))
+          continue;
+        worklist.pushIfNotVisited(successor);
+        consumedBlocks.insert(successor);
+      }
+    }
+  }
+
+  // Third, find the blocks on the boundary between the originally-live blocks
+  // and the originally-live-but-consumed blocks.  These are new boundary
+  // blocks.
+  for (auto *block : consumedBlocks) {
+    for (auto *predecessor : block->getPredecessorBlocks()) {
+      if (!consumedBlocks.contains(predecessor) &&
+          !finalConsumingBlocks.contains(predecessor)) {
+        // Add "the instruction(s) before the terminator" of the predecessor to
+        // liveness.
+        if (auto *inst =
+                predecessor->getTerminator()->getPreviousInstruction()) {
+          liveness.updateForUse(inst, /*lifetimeEnding*/ false);
+        } else {
+          for (auto *grandPredecessor : predecessor->getPredecessorBlocks()) {
+            liveness.updateForUse(grandPredecessor->getTerminator(),
+                                  /*lifetimeEnding*/ false);
```

Comment:
Can you try out a test case where `predecessor` is itself a control-flow merge? We currently assume that a phi operand (branch instruction) can never be the last use unless it is lifetime-ending (This needs to be more clearly commented and asserted somewhere in PrunedLiveness). Breaking this assumption probably means we end up visiting the same insertion point multiple times.

One option is to simply insert the destroy_value at the correct place here and make it the live use.

Another option is to add a special case to visitInsertionPoints to handle nonconsuming-last-use-phis with extra book-keeping.

Reply:
Oops, it looks like all the tests I added didn't make it into the PR.  Fixed that (copy_propagation_onone).

Reply:
Added.  It exposed an issue when attempting to extend a boundary along multiple cfg edges with the same destination.